backend/execution_cache/src/admission.go

```diff
@@ -0,0 +1,143 @@
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+type OperationType string
+
+// patchOperation is an operation of a JSON patch, see https://tools.ietf.org/html/rfc6902 .
+type patchOperation struct {
+	Op    OperationType `json:"op"`
+	Path  string        `json:"path"`
+	Value interface{}   `json:"value,omitempty"`
+}
+
+// admitFunc is a callback for admission controller logic. Given an AdmissionRequest, it returns the sequence of patch
+// operations to be applied in case of success, or the error that will be shown when the operation is rejected.
+type admitFunc func(*v1beta1.AdmissionRequest) ([]patchOperation, error)
+
+const (
+	ContentType     string = "Content-Type"
+	JsonContentType string = "application/json"
+)
+
+var (
+	universalDeserializer = serializer.NewCodecFactory(runtime.NewScheme()).UniversalDeserializer()
+)
+
+// isKubeNamespace checks if the given namespace is a Kubernetes-owned namespace.
+func isKubeNamespace(ns string) bool {
+	return ns == metav1.NamespacePublic || ns == metav1.NamespaceSystem
+}
+
+// doServeAdmitFunc parses the HTTP request for an admission controller webhook, and -- in case of a well-formed
+// request -- delegates the admission control logic to the given admitFunc. The response body is then returned as raw
+// bytes.
+func doServeAdmitFunc(w http.ResponseWriter, r *http.Request, admit admitFunc) ([]byte, error) {
+	// Step 1: Request validation. Only handle POST requests with a body and json content type.
+
+	if r.Method != http.MethodPost {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return nil, fmt.Errorf("invalid method %q, only POST requests are allowed", r.Method)
+	}
+
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, fmt.Errorf("could not read request body: %v", err)
+	}
+
+	if contentType := r.Header.Get(ContentType); contentType != JsonContentType {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, fmt.Errorf("unsupported content type %q, only %q is supported", contentType, JsonContentType)
+	}
+
+	// Step 2: Parse the AdmissionReview request.
+
+	var admissionReviewReq v1beta1.AdmissionReview
+
+	if _, _, err := universalDeserializer.Decode(body, nil, &admissionReviewReq); err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, fmt.Errorf("could not deserialize request: %v", err)
+	} else if admissionReviewReq.Request == nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return nil, errors.New("malformed admission review: request is nil")
+	}
+
+	// Step 3: Construct the AdmissionReview response.
+
+	admissionReviewResponse := v1beta1.AdmissionReview{
+		Response: &v1beta1.AdmissionResponse{
+			UID: admissionReviewReq.Request.UID,
+		},
+	}
+
+	var patchOps []patchOperation
+	// Apply the admit() function only for non-Kubernetes namespaces. For objects in Kubernetes namespaces, return
+	// an empty set of patch operations.
+	if !isKubeNamespace(admissionReviewReq.Request.Namespace) {
```

Comment:
No strong opinion, but I'd vote the following for better code readability
```go
if isKubeNamespace(admissionReviewReq.Request.Namespace) {
        return allowedResponse(admissionReviewReq.Request.UID, nil), nil
}

var patchOps []patchOperation

patchOps, err = admit(admissionReviewReq.Request)
if err != nil {
        return errorResponse(admissionReviewReq.Request.UID, err), nil
}

return allowedResponse(admissionReviewReq.Request.UID, patchBytes), nil
```
It also feels strange setting the `err` variable inside an `if` and checking it outside of it (when `!isKubeNamespace` is true).

Reply:
Thanks! Refactored this part.